### PR TITLE
Centralize requests, replace "updates" with one "patch"

### PR DESCRIPTION
**Pull request by alexander-demichev**
Centralize requests, replace "updates" with one "patch". Instead of doing a lot of deep copies and updates, do patch only one time.

### CODE REVIEW

**Comment by enxebre** [comment]
Can we please add a PR description with the motivation behind this PR?
can we please put the go mod changes into its own commit?

File: pkg/actuators/machine/actuator.go [hunk 1]
```diff
@@ -654,3 +651,18 @@ func getClusterID(machine *machinev1.Machine) (string, bool) {
 	}
 	return clusterID, ok
 }
+
+func (a *Actuator) patchMachine(ctx context.Context, machine *machinev1.Machine, machineToBePatched client.Patch) error {
+	// Patch machine
+	if err := a.client.Patch(context.Background(), machine, machineToBePatched); err != nil {
```
Comment (enxebre) [inline pkg/actuators/machine/actuator.go]: this should be using the ctx you passing in the header

File: pkg/actuators/machine/actuator.go [hunk 2]
```diff
@@ -654,3 +651,18 @@ func getClusterID(machine *machinev1.Machine) (string, bool) {
...
+		klog.Errorf("Failed to update machine %q: %v", machine.GetName(), err)
+		return err
+	}
+
+	//Patch status
+	if err := a.client.Status().Patch(context.Background(), machine, machineToBePatched); err != nil {
```
Comment (enxebre) [inline pkg/actuators/machine/actuator.go]: this should be using the ctx you passing in the header

File: pkg/actuators/machine/actuator.go [hunk 3]
```diff
@@ -612,27 +601,35 @@ func (a *Actuator) updateStatus(machine *machinev1.Machine, instance *ec2.Instan
 	} else {
 		awsStatus.InstanceID = instance.InstanceId
 		awsStatus.InstanceState = instance.State.Name
-
-		addresses, err := extractNodeAddresses(instance)
-		if err != nil {
-			glog.Errorf("%s: Error extracting instance IP addresses: %v", machine.Name, err)
-			return err
+		if instance.PublicIpAddress != nil {
+			networkAddresses = append(networkAddresses, corev1.NodeAddress{
```
Comment (enxebre) [inline pkg/actuators/machine/actuator.go]: why is this dropping `extractNodeAddresses`, we want to use it here

File: pkg/actuators/machine/actuator.go [hunk 4]
```diff
@@ -641,7 +638,7 @@ func (a *Actuator) updateStatus(machine *machinev1.Machine, instance *ec2.Instan
 	// we get a public IP populated more quickly.
 	if awsStatus.InstanceState != nil && *awsStatus.InstanceState == ec2.InstanceStateNamePending {
```
Comment (enxebre) [inline pkg/actuators/machine/actuator.go]: we should probably move this check up outside of this function, so the func does not error, so the patch request takes place so the machine and the status are updated nevertheless

File: pkg/actuators/machine/actuator.go [hunk 5]
```diff
@@ -24,13 +24,12 @@ import (
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/service/ec2"
 	"github.com/golang/glog"
+	clustererror "github.com/openshift/cluster-api/pkg/controller/error"
```
Comment (enxebre) [inline pkg/actuators/machine/actuator.go]: openshift/cluster-api is deprecated
Please use anything from machine-api-operator

Comment (enxebre) [inline pkg/actuators/machine/actuator.go]: this seems to be added my mistake to the commit, can we remove it?

Comment (enxebre) [reply pkg/actuators/machine/actuator.go]: why do we need a revendor commit at all?

File: pkg/actuators/machine/actuator.go [hunk 6]
```diff
@@ -641,7 +646,7 @@ func (a *Actuator) updateStatus(machine *machinev1.Machine, instance *ec2.Instan
 	// we get a public IP populated more quickly.
 	if awsStatus.InstanceState != nil && *awsStatus.InstanceState == ec2.InstanceStateNamePending {
 		glog.Infof("%s: Instance state still pending, returning an error to requeue", machine.Name)
-		return &machinecontroller.RequeueAfterError{RequeueAfter: requeueAfterSeconds * time.Second}
+		return &clustererror.RequeueAfterError{RequeueAfter: requeueAfterSeconds * time.Second}
```
Comment (enxebre) [inline pkg/actuators/machine/actuator.go]: why this change?

File: pkg/actuators/machine/actuator.go [hunk 7]
```diff
@@ -471,6 +476,11 @@ func (a *Actuator) Update(context context.Context, machine *machinev1.Machine) e
 		return fmt.Errorf("%s: failed to patch machine: %v", machine.Name, err)
 	}
 
+	err = a.requeueIfInstancePending(machine)
```
Comment (enxebre) [inline pkg/actuators/machine/actuator.go]: this can be simplified by just  `a.requeueIfInstancePending(machine)`, no need for `if` statement

File: pkg/actuators/machine/actuator.go [hunk 8]
```diff
@@ -140,6 +140,11 @@ func (a *Actuator) Create(context context.Context, machine *machinev1.Machine) e
...
```
Comment (enxebre) [inline pkg/actuators/machine/actuator.go]: this can be simplified by just  `a.requeueIfInstancePending(machine)`, no need for `if` statement

File: pkg/actuators/machine/actuator.go [hunk 9]
```diff
@@ -140,7 +140,7 @@ func (a *Actuator) Create(context context.Context, machine *machinev1.Machine) e
 		return fmt.Errorf("%s: failed to patch machine: %v", machine.Name, err)
 	}
 
-	return nil
+	return a.requeueIfInstancePending(machine)
```
Comment (enxebre) [inline pkg/actuators/machine/actuator.go]: we might want to pass `newestInstance` here instead as that's the source of truth

**Comment by enxebre** [comment]
@bison PTAL

File: pkg/actuators/machine/actuator.go [hunk 10]
```diff
@@ -105,140 +103,121 @@ func (a *Actuator) handleMachineError(machine *machinev1.Machine, err error, eve
...
-	}
-
-	return machineCopy, nil
+	return nil
 }
 
-// updateProviderID adds providerID in the machine spec
-func (a *Actuator) updateProviderID(machine *machinev1.Machine, instance *ec2.Instance) (*machinev1.Machine, error) {
+// setProviderID adds providerID in the machine spec
+func (a *Actuator) setProviderID(machine *machinev1.Machine, instance *ec2.Instance) error {
```
Comment (bison) [inline pkg/actuators/machine/actuator.go]: This method doesn't seem to ever actually return an error. Seems like we could just drop that return type. Though, if we want to keep it as-is just to have a common API among setter methods, that's fine.

**Comment by alexander-demichev** [comment]
@bison Good catch, I've just added a commit that replaces `fmt.Errorf()` with `handleMachineError()`

File: pkg/actuators/machine/actuator.go (hunk 3 above)
Comment (bison) [reply pkg/actuators/machine/actuator.go]: This was marked as resolved with no response.

I don't understand this bit. The call to `extractNodeAddresses` already sets all these. Is this needed? Wouldn't it duplicate all the node addresses?
